src/ws/Socket.ts

```diff
@@ -461,16 +462,17 @@ export class Socket extends EventEmitter {
 
         // If we already authed, it means we're reconnecting and should
         // establish authentication again.
-        if (this._authpacket) {
+        if (this._authpacket || this._optOutEventsArgs.length) {
             // tslint:disable-next-line no-floating-promises
-            this.call(authMethod, this._authpacket, { force: true })
+            const promise = this._optOutEventsArgs.length
+                ? this.call('optOutEvents', this._optOutEventsArgs, { force: true }).catch(() =>
+                      this.emit('error', new UnknownCodeError()),
+                  )
+                : Promise.resolve();
+            promise
+                .then(() => this.emit('optOutResult'))
+                .then(() => this.call(authMethod, this._authpacket, { force: true }))
```

Comment:
This code could be refactored to be a lot easier to follow, essentially as poor man's async/await 😛 

```js
let promise = Promose.resolve();
if (this._authpacket) {
  promise = promise.then(() => this._doInitialAuth());
}
if (this. _optOutEventsArgs.length) {
  promise = promise.then(() => this. _doInitialOptOutCall());
}

promise.then(bang).catch(err => /* catch authentication failed errors */);
```

Reply:
You can also use async/await, we transpile to es5 anyways.

Reply:
Async await is huge and slow in downlevelled code

Reply:
@connor4312 I can refactor that but that would make `bang`  function asynchronous(I'm pretty sure it'll break some of the unit tests too) and not sure about consequential impact on clients using because of this behavior. I would rather merge this PR for now as the code works as expected. Also, talked with @ProbablePrime he's expecting someone from his team to refactor this code anyway.

Reply:
That should be fine. In the case where there's an auth packet, the `bang` function is already called asynchronously. In the other case, deferring it until the end of the current event loop tick (a `then` on a resolved promise) should have no meaningful impact.